### PR TITLE
Everloop example for ESP32

### PR DESCRIPTION
**Pull request by jeikabu**
- Build/deploy instructions in README.md
- Pretty much a straight port of esp32-platformio and matrixio_hal_esp32 everloop code
- Initial MatrixBus impl using ESP-IDF

### CODE REVIEW

**Comment by jeikabu** [comment]
Couple of things worth mentioning:
- `MatrixBus::write()` and `read()` currently assume the "address" is stored in the first 4 bytes (followed by the data size) of the buffer.  I assume this made sense on Linux, but it's awkward on the ESP32 because I have to pull it out because the way it's used is a bit different.  Perhaps we could change that, or add additional methods like `write_address()`.
- `transmute::<&mut [i32], &mut [u8]>(&mut request)` truncates the data.  Meaning if `request = [0u32; 1]`, transmute returns a slice of the same length so you get a single `u8` instead of 4 bytes like you expect.
- I haven't dug very deep into using SPI peripherals with ESP-IDF, it's currently mostly the same as what was in matrixio_hal_esp32 except I replaced the 4KB global buffers with 512B stack buffers.  Now that I have it working I'll investigate it more closely.

**Comment by Hermitter** [comment]
Thanks for the work on the ESP32 bus implementation!

I'm open to adding convenience functions for reading/writing the address. I haven't seen how it's handled for the ESP32 so I'll need to look more into it.

For the transmute truncation issue, we can probably use the [ByteOrder crate](https://crates.io/crates/byteorder) as an alternative (supports no_std). 

```toml
[dependencies]
byteorder = { version = "1.34", default-features = false }
```

```rust
use byteorder::{ByteOrder, LittleEndian};

fn main() {
    const BUFFER_SIZE: usize = 16;
    let mut buffer: [u8; BUFFER_SIZE] = [0; BUFFER_SIZE];

    let request: [i32; BUFFER_SIZE / 4] = [1000, 2000, 3000, 4000];
    LittleEndian::write_i32_into(&request, &mut buffer);
    
    assert_eq!(
        buffer,
        [232, 3, 0, 0, 208, 7, 0, 0, 184, 11, 0, 0, 160, 15, 0, 0]
    );
}
```

I should have more time next week to review your PR and tinker with these new changes.

**Comment by jeikabu** [comment]
If we remove "address" and "length" from the front of the buffers we don't need to bother with the endianness or using `transmute` to "cast" because it's just raw bytes.  If the data contains values larger than 8 bits they'll obviously need to serialize it appropriately.

Basically, change:
```rust
trait MatrixBus {
    fn write(&self, write_buffer: &mut [u8]);
    fn read(&self, read_buffer: &mut [u8]);
}
```
Into:
```rust
trait MatrixBus {
    fn write(&self, address: u16, write_buffer: & [u8]);
    fn read(&self, address: u16, read_buffer: &mut [u8]);
}
```

The various `MatrixBus` impls are then responsible for packing everything in the correct format, alignment, etc.

File: src/everloop/mod.rs [hunk 1]
```diff
@@ -26,37 +28,34 @@ impl<'a> Everloop<'a> {
...
-        request.push((device_leds * 4) as i32).unwrap(); // each LED RGBW requires 4 bytes
+        let led_bytes = device_leds * core::mem::size_of::<LedBytes>();
+        let mut request: Vec<u8, MAX_LED_BYTES> = Vec::new();
+        request.extend_from_slice(&(fpga_address::EVERLOOP as i32).to_ne_bytes()).unwrap();
+        request.extend_from_slice(&(led_bytes as i32).to_ne_bytes()).unwrap(); // each LED RGBW requires 4 bytes
 
         // store all LED colors given
         for led in leds {
-            request.push(led.as_bytes()).unwrap();
+            request.extend_from_slice(&led.to_bytes()).unwrap();
```
Comment (Hermitter) [inline src/everloop/mod.rs:67]: This throws an error when I run it on the Raspberry Pi. 

Example Code:
```
    let bus = hal::bus::init();
    let everloop = hal::Everloop::new(&*bus);
    everloop.set_all(hal::Rgbw::black());
```

Error Log:
```
thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: ()', src/everloop/mod.rs:44:13
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

**Review by Hermitter — COMMENTED** [review]
So far, the only breaking change seems to be with the Everloop in an `std` environment. @samreenislam was able to successfully compile and test your ESP32 example!

**Comment by Hermitter** [comment]
> If we remove "address" and "length" from the front of the buffers we don't need to bother with the endianness or using `transmute` to "cast" because it's just raw bytes. If the data contains values larger than 8 bits they'll obviously need to serialize it appropriately.
> 
> Basically, change:
> 
> ```rust
> trait MatrixBus {
>     fn write(&self, write_buffer: &mut [u8]);
>     fn read(&self, read_buffer: &mut [u8]);
> }
> ```
> 
> Into:
> 
> ```rust
> trait MatrixBus {
>     fn write(&self, address: u16, write_buffer: & [u8]);
>     fn read(&self, address: u16, read_buffer: &mut [u8]);
> }
> ```
> 
> The various `MatrixBus` impls are then responsible for packing everything in the correct format, alignment, etc.

Looks cleaner! Once the PR is merged, I think this signature would be better to have.

File: src/everloop/mod.rs (hunk 1 above)
Comment (jeikabu) [reply src/everloop/mod.rs:67]: With a Creator, right?  That buffer isn't large enough for the extra header bytes.

Comment (Hermitter) [reply src/everloop/mod.rs:67]: Yep! I haven't tested it with the Voice yet.

**Comment by jeikabu** [comment]
It works with Voice and I don't have a Creator to test with, but this should fix that assert.

**Comment by Hermitter** [comment]
Sorry for the delay! Was able to test that the changes work on both Voice and Creator.

**Pull request merged**